src/App.js

```diff
@@ -1,65 +1,163 @@
+        edited: false
+      },
+      {
+        id: 3,
+        caption: 'Three',
+        text: 'third',
+        checked: false,
+        edited: false
+      },
+      {
+        id: 4,
+        caption: 'Four',
+        text: 'fourth',
+        checked: false,
+        edited: false
+      },
+      {
+        id: 5,
+        caption: 'Five',
+        text: 'Fifth',
+        checked: false,
+        edited: false
+      },
+      {
+        id: 6,
+        caption: 'Six',
+        text: 'sixth',
+        checked: false,
+        edited: false
+      },
+      {
+        id: 7,
+        caption: 'Seven',
+        text: 'seventh',
+        checked: false,
+        edited: false
+      },
+      {
+        id: 8,
         caption: 'Caption',
         text: 'Text...',
-        checked: false
-    },
-    edited: false
+        checked: false,
+        edited: false
+      }
+    ],
+    checked: false
   });
 
-  function checkboxChange() {
+  function checkboxChange(id) {
+    const cardIndex = cardState.cards.findIndex(c => {
+      return c.id === id;
+    });
+
+    const card = {...cardState.cards[cardIndex]};
+    card.checked = !card.checked;
+
+    const cards = [...cardState.cards];
+    cards[cardIndex] = card;
     setCardState({
       ...cardState,
-      card: {
-        ...cardState.card,
-        checked: !cardState.card.checked
-      }
+      cards: cards
     });
   }
 
-  function editMode() {
+  function editMode(id) {
+    const cardIndex= cardState.cards.findIndex(c => {
+      return c.id === id;
+    });
+    const card = {...cardState.cards[cardIndex]};
+    card.checked = false;
+    card.edited = true;
+
+    const cards = [...cardState.cards];
+    cards[cardIndex] = card;
     setCardState({
-      card: {
-        ...cardState.card,
-        checked: false
-      },
-      edited: true
+      ...cardState,
+      cards: cards
     });
   }
 
-  function saveEditing(x, y) {
+  function saveEditing(arr, id) {
+    const cardIndex= cardState.cards.findIndex(c => {
+      return c.id === id;
+    });
+    const card = {...arr};
+    card.checked = false;
+
+    const cards = [...cardState.cards];
+    cards[cardIndex] = card;
+
     setCardState({
-      card: {
-        ...cardState.card,
-        caption: x,
-        text: y
-      },
-      edited: false
+      ...cardState,
+      cards: cards
     });
   }
 
-  function cancelEditing() {
+  function cancelEditing(id) {
+    const cardIndex = cardState.cards.findIndex(c => {
+      return c.id === id;
+    });
+    const card = {...cardState.cards[cardIndex]};
+    card.edited = false;
+
+    const cards = [...cardState.cards];
+    cards[cardIndex] = card;
+
     setCardState({
       ...cardState,
-      edited: false
+      cards: cards
+    });
+  }
+
+  function mainChecked() {
+    cardState.cards.forEach(c => {
+      return c.edited = false;
+    });
+    setCardState({
+      checked: !cardState.checked,
+      cards: cardState.cards
```

Comment:
You're setting self values. If you need to let hook know that value should not be changed - use spread operator to copy all the existing properties, then define the ones to overwrite.